decisions/2024-04-15_decimal_pack_size.md

```diff
@@ -0,0 +1,131 @@
+- Data reviewed on remote sites will not match the same data viewed on central server.
+- Might need explaining/complex mappings in integrations.
+
+### Option 3 - Can convert units?
+
+Specifically for cases like mL and L, it might be handy in repacks to allow 1L -> 1000mL. This could be converted. Then instead of items having units, they have a unit type, i.e. _volume_, _mass_ or _each_. Then when sending a fractional amount, it could automatically repack into a different volume. Monetary values may need to remain relative to a recorded base unit to avoid doing conversions of that as well, or perhaps we convert units too though they may become very small values ($1/L -> $0.001/mL).
+
+In translation fractional pack sizes could be converted from one unit to another smaller unit that results in a whole number, but only if they're defined centrally. Unfortunately we can't universally predict this - as Per the SAP example a user might user "case" and "each", where there are 6 "each" in one "case". Do we define a separate "each" for each possible ration? 1:6, 1:10, 1:12... we certainly shouldn't generate units to get around it!
+
+Pros:
+
+- Tidy and intuitive-ish specifically for L and mL. For pharmacy users it means data stays directly mapping with physical stock.
+- Fine for integrations?
+
+Cons:
+
+- Wide ranging impact and complexity especially in translations.
+- Units are currently controlled in Legacy, so all would have to be preconfigured before integration is attempted on remote site.
+- Doesn't logically work for all units, so feasibility is near 0 IMO.
+
+### Option 4 - Switch all f64 amount fields to a monetary safe type
+
+Decimals would be safe then. pgsql has one of these built in, sqlite does not. We'd need a rust implementation to manage these monetary values. One such rust crate is [rusty-money](https://crates.io/crates/rusty-money). It is of note that rust is a young language, and such libraries are relatively immature compared to [Java](https://docs.oracle.com/cd/E13166_01/alcs/docs51/javadoc/com/elasticpath/domain/misc/Money.html), [.NET](https://learn.microsoft.com/en-us/dotnet/fundamentals/runtime-libraries/system-decimal) etc.
+
+(Note .NET's `Decimal` has basically the same problems as f64 still, but the errors are relatively much smaller! Is a fancy 128bit number implementation rather than f64 standard)
+
+This is too big a task. All reports and all database and business logic would have to be rewritten. OMS needed to commit to this early if at all so I think the ship has sailed, won't write anymore about it! Would have incurred a complexity tax on all development anyway, and Legacy would still be sending us f64 to deal with lol.
+
+## Decision
+
+**Option 1 - it is the simplest and the consequences are generally known.**
```

Comment:
Ok, agree, mainly because, less logic in translation = less use case. And also if we don't spend the extra time trying to account for repeating numbers (which I think have a very very small chance of happening). And also we use pack variants to help use here:

Like your previous example: `1 pack of 1L is repacked to 40 packs of 0.05L`, we create new pack variant = 0.05L, now instead of showing `0.05L` in pack size, it would say the short name "5ml bottle" x 40.  <- and this can work without repack, user can choose to repack or use the pack variant 'lens'

Reply:
> Ok, agree, mainly because, less logic in translation = less use case. And also if we don't spend the extra time trying to account for repeating numbers (which I think have a very very small chance of happening). And also we use pack variants to help use here:
> 
> Like your previous example: `1 pack of 1L is repacked to 40 packs of 0.05L`, we create new pack variant = 0.05L, now instead of showing `0.05L` in pack size, it would say the short name "5ml bottle" x 40. <- and this can work without repack, user can choose to repack or use the pack variant 'lens'

Pack variants are currently `i32`, am I misunderstanding or would we need to change that to `f64` too?

Reply:
yes